File: src/main/java/com/baidu/algorithm/slidewindow/_239_Sliding_Window_Maximum.py
from collections import deque


class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        dq = deque()
        res = []
        for i, num in enumerate(nums):
            if dq and dq[0] == i - k:
                dq.popleft()

            while dq and nums[dq[-1]] <= num:
                dq.pop()

            dq.append(i)
            if i >= k - 1:
                res.append(nums[dq[0]])

        return res

File: src/main/java/com/baidu/algorithm/slidewindow/_239_Sliding_Window_Maximum.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.slidewindow;

import java.util.Deque;
import java.util.LinkedList;

/**
 * _239_Sliding_Window_Maximum
 *
 * @author xuhaoran01
 */
public class _239_Sliding_Window_Maximum {

    public int[] maxSlidingWindow(int[] nums, int k) {

        if (nums == null || nums.length == 0 || k > nums.length) {
            return new int[0];
        }

        int[] res = new int[nums.length - k + 1];
        Deque<Integer> deque = new LinkedList<>();

        for (int i = 0; i < nums.length; i++) {

            if (!deque.isEmpty() && deque.peekLast() == i - k) {
                deque.removeLast();
            }

            while (!deque.isEmpty() && nums[deque.peekFirst()] <= nums[i]) {
                deque.removeFirst();
            }

            deque.addFirst(i);
            if (i >= k - 1) {
                res[i - k + 1] = nums[deque.peekLast()];
            }
        }

        return res;
    }
}
